File: tests/e2e-ft-hint-contract.test.mjs
/**
 * tests/e2e-ft-hint-contract.test.mjs — FT-A1 sealed-offer hint, RULE 30.
 *
 * ONE vector file, `tests/e2e-ft-hint-vectors.json`, consumed by TWO
 * implementations:
 *
 *   - this file — the WEB producer (`ftHintFor`, lib/fileTransfer/frames.ts, on
 *     the seal path of `sealOutbound` in hooks/useE2e.ts) fed to the REAL relay
 *     accessor `ftOfferMetadata`, extracted from server.js by the same
 *     bracket-balanced slice tests/ft-relay.test.mjs uses;
 *   - dnkdialer-android/app/src/test/java/com/dnkdialer/companion/
 *     E2eFtHintContractTest.kt — the PHONE producer (`FileTransfer.hintFor` +
 *     `E2eFrameGate.attachHint`) over the SAME rows.
 *
 * ## The defect this exists for
 *
 * server.js:2361-2392 fails CLOSED when a sealed FILE_OFFER carries no
 * plaintext `ft.id` hint. The phone attached one (E2eFrameGate.kt:182); the
 * browser sealed FILE_OFFER as a BARE envelope and special-cased only
 * CALL_STATUS. So EVERY browser->phone file on ANY sealed pair — including an
 * unverified 0/0 one — was refused `bad_hint`: Dennis's "0 bytes, no progress
 * bar", proven on PROD 6d0aa98 by ACCEPT-9.
 *
 * ## Why the relay half is EXTRACTED and not re-typed
 *
 * A contract test that restates the relay's rules in its own JavaScript proves
 * that two copies of one author's opinion agree. The rules here are read out of
 * server.js itself, so a relay-side change to the hint gate that the web
 * producer does not follow turns this file red.
 *
 * Run: node tests/e2e-ft-hint-contract.test.mjs
 */

import { createRequire } from 'node:module';
import { readFileSync, existsSync } from 'node:fs';
import path from 'node:path';

import { ftHintFor, FT_HINT_KEY } from '../lib/fileTransfer/frames.ts';

const require = createRequire(import.meta.url);
const VECTORS = require('./e2e-ft-hint-vectors.json');
const ROOT = path.resolve(import.meta.dirname, '..');

let passed = 0;
let failed = 0;
function check(name, ok, detail = '') {
  if (ok) { passed++; return true; }
  failed++;
  console.error(`  FAIL  ${name}${detail ? ` — ${detail}` : ''}`);
  return false;
}
const eq = (name, got, want) =>
  check(name, got === want, `got ${JSON.stringify(got)} want ${JSON.stringify(want)}`);

/** CRLF-safe (.gitattributes pins lib/hooks to LF; a fresh checkout may not be). */
function stripComments(src) {
  return src
    .replace(/\r\n/g, '\n')
    .replace(/\/\*[\s\S]*?\*\//g, '')
    .replace(/^[ \t]*\/\/.*$/gm, '')
    .replace(/[ \t]\/\/.*$/gm, '');
}

// ── 0. the vector file is not empty and not silently re-versioned ──────────
eq('vectors: version', VECTORS.version, 1);
eq('vectors: the hint key both surfaces splice', VECTORS.hintKey, FT_HINT_KEY);
check('vectors: the rows survived (a vectors suite with no rows passes vacuously)',
  Array.isArray(VECTORS.rows) && VECTORS.rows.length >= 10, String(VECTORS.rows?.length));
check('vectors: an accepted BROWSER offer row exists — the defect row',
  VECTORS.rows.some((r) => r.producer === 'web' && r.relay?.ok === true));
{
  // The ONE recorded cross-surface divergence (hint-id-uppercase): the web
  // producer validates the id against the relay's pattern, FileTransfer.hintFor
  // does not. Both land on the same relay verdict, so it is a documented
  // difference and not a defect — but it must stay DOCUMENTED and it must stay
  // ONE row, or "the two surfaces emit one shape" quietly stops being true.
  const diverging = VECTORS.rows.filter((r) => r.expectHintPhone !== undefined);
  eq('vectors: exactly one row records a cross-surface divergence', diverging.length, 1);
  eq('vectors: and it is the malformed-id row', diverging[0]?.id, 'hint-id-uppercase');
  eq('vectors: the web side of that row emits NO hint', diverging[0]?.expectHint, null);
  check('vectors: the divergence is explained in the row label, not left bare',
    /divergence/i.test(diverging[0]?.label ?? ''));
  eq('vectors: and both surfaces still end at the same relay verdict',
    diverging[0]?.relay?.drop, 'bad_hint');
}
check('vectors: the pre-fix no-hint regression row exists',
  VECTORS.rows.some((r) => r.wireOmitsHint === true && r.relay?.drop === 'bad_hint'
    && r.envelope[FT_HINT_KEY] === undefined && r.expectHint !== null));

// ── 1. the relay's REAL hint gate, sliced out of server.js ────────────────
const SERVER_SRC = readFileSync(path.join(ROOT, 'server.js'), 'utf8');

/** Bracket-balanced `function NAME(` extraction (tests/ft-relay.test.mjs idiom). */
function extractFn(name) {
  const re = new RegExp(`(?:^|\\n)\\s*(?:async\\s+)?function ${name}\\s*\\(`);
  const m = re.exec(SERVER_SRC);
  if (!m) throw new Error(`function ${name} not found in server.js`);
  const start = SERVER_SRC.indexOf('function', m.index);
  let depth = 0;
  for (let j = SERVER_SRC.indexOf('{', start); j < SERVER_SRC.length; j++) {
    if (SERVER_SRC[j] === '{') depth++;
    else if (SERVER_SRC[j] === '}' && --depth === 0) return SERVER_SRC.slice(start, j + 1);
  }
  throw new Error(`unterminated ${name}`);
}
function extractConst(name) {
  const re = new RegExp(`(?:^|\\n)\\s*const ${name} =`);
  const m = re.exec(SERVER_SRC);
  if (!m) throw new Error(`const ${name} not found in server.js`);
  const start = SERVER_SRC.indexOf('const ', m.index);
  let depth = 0;
  for (let j = start; j < SERVER_SRC.length; j++) {
    const c = SERVER_SRC[j];
    if ('([{'.includes(c)) depth++;
    else if (')]}'.includes(c)) depth--;
    else if (c === ';' && depth === 0) return SERVER_SRC.slice(start, j + 1);
  }
  throw new Error(`unterminated const ${name}`);
}

const relay = new Function([
  extractConst('FT_MAX_FILE_BYTES'),
  extractFn('ftOfferMetadata'),
  extractFn('ftFrameId'),
  'return { ftOfferMetadata, ftFrameId };',
].join('\n\n'))();

check('the relay accessor really came from server.js (not a local re-type)',
  typeof relay.ftOfferMetadata === 'function' && typeof relay.ftFrameId === 'function');

// ── 2. every row: producer -> wire envelope -> relay verdict ──────────────
console.log(`FT hint contract — ${VECTORS.rows.length} rows from tests/e2e-ft-hint-vectors.json`);
for (const row of VECTORS.rows) {
  // (a) the WEB producer. `both`/`phone` rows are asserted here too: the web
  //     side must reach the SAME hint for a phone-authored body, which is the
  //     only way "the two surfaces emit one shape" is actually measured.
  const hint = ftHintFor(row.type, row.body);
  if (row.expectHint === null) {
    eq(`${row.id}: web producer emits NO hint`, hint, null);
  } else {
    check(`${row.id}: web producer emits a hint`, hint !== null);
    eq(`${row.id}: web hint.id`, hint?.id, row.expectHint.id);
    eq(`${row.id}: web hint.size`, hint?.size, row.expectHint.size);
  }

  // (b) the wire envelope the vector records is what that producer would build:
  //     the four authenticated fields, plus the hint as a SIBLING (never inside
  //     the ciphertext, never in the AAD).
  const wireHint = row.envelope[FT_HINT_KEY];
  if (row.wireOmitsHint) {
    // The REGRESSION row: a body the producer must hint, recorded on the wire
    // WITHOUT one. It is the pre-fix shape, so it is asserted from both ends —
    // the producer emits a hint for this very body (above), and the recorded
    // envelope is missing it (here). Those two cannot both be true of shipped
    // code, which is exactly what makes this row the defect.
    check(`${row.id}: the recorded envelope is the pre-fix shape (no ft sibling)`,
      wireHint === undefined, JSON.stringify(wireHint));
  } else if (row.expectHint === null) {
    check(`${row.id}: the recorded envelope's hint is absent or is the malformed one under test`,
      wireHint === undefined || row.relay?.drop === 'bad_hint');
  } else {
    check(`${row.id}: the recorded envelope carries the producer's hint verbatim`,
      wireHint && wireHint.id === hint.id && wireHint.size === hint.size,
      JSON.stringify(wireHint));
  }
  check(`${row.id}: the envelope keeps the four authenticated fields`,
    row.envelope.e !== undefined && typeof row.envelope.kid === 'string'
    && typeof row.envelope.s === 'number' && typeof row.envelope.c === 'string');
  check(`${row.id}: nothing but {e,kid,s,c${wireHint ? ',ft' : ''}} leaves in the clear`,
    Object.keys(row.envelope).every((k) => ['e', 'kid', 's', 'c', FT_HINT_KEY].includes(k)),
    Object.keys(row.envelope).join(','));
  for (const leak of ['name', 'mime', 'sha256', 'from']) {
    check(`${row.id}: ${leak} stays sealed`, !(leak in row.envelope));
  }

  // (c) the relay's verdict on that exact envelope.
  if (row.relay === null) {
    // A non-offer frame is routed by TYPE. Assert it the way the relay does:
    // no readable id at all, which is what makes TYPE routing the only option.
    eq(`${row.id}: a sealed non-offer carries no readable id`, relay.ftFrameId(row.envelope), null);
    continue;
  }
  const meta = relay.ftOfferMetadata(row.envelope);
  eq(`${row.id}: relay ok`, meta.ok, row.relay.ok);
  eq(`${row.id}: relay sees it as sealed`, meta.sealed, row.relay.sealed);
  eq(`${row.id}: relay id`, meta.id, row.relay.id);
  eq(`${row.id}: relay size`, meta.size, row.relay.size);
  check(`${row.id}: relay never hands back an account`,
    !('userId' in meta) && !('account' in meta));
  // The drop reason the relay would count for this frame, taken from the ONE
  // expression in server.js that decides it (asserted verbatim in step 3).
  const drop = meta.ok ? null : (meta.sealed ? 'bad_hint' : 'malformed');
  eq(`${row.id}: drop reason`, drop, row.relay.drop);
}

// ── 3. the product path — pins, so this suite measures the SHIPPED code ───
{
  const hook = stripComments(readFileSync(path.join(ROOT, 'hooks', 'useE2e.ts'), 'utf8'));
  check('useE2e imports the ONE hint producer (no second local copy)',
    /import \{[^}]*ftHintFor[^}]*\} from '@\/lib\/fileTransfer\/frames\.ts';/.test(hook));
  const fn = hook.slice(hook.indexOf('const sealOutbound'));
  const body = fn.slice(0, fn.indexOf('\n  }, ['));
  check('the sealOutbound body actually sliced (not an empty fragment)',
    body.includes('session.seal') && body.length > 200, String(body.length));
  check('sealOutbound attaches the hint on the seal path',
    /ftHintFor\(type, payload as Record<string, unknown>\)/.test(body), body);
  check('the hint rides as a SIBLING of the envelope, not inside it',
    /return ft \? \{ \.\.\.env, ft \} : env;/.test(body), body);
  check('CALL_STATUS keeps its own field split (this lane changed nothing there)',
    /splitCallStatus\(payload as Record<string, unknown>\)/.test(body));

  const server = stripComments(SERVER_SRC);
  check('server.js still counts a sealed refusal as bad_hint (the reason this suite reads)',
    /ftCountDrop\(token, 'FILE_OFFER', meta\.sealed \? 'bad_hint' : 'malformed'\);/.test(server));

  const kt = path.join(ROOT, 'dnkdialer-android', 'app', 'src', 'test', 'java', 'com',
    'dnkdialer', 'companion', 'E2eFtHintContractTest.kt');
  check('the Kotlin twin of this contract exists', existsSync(kt), kt);
  const ktSrc = existsSync(kt) ? readFileSync(kt, 'utf8') : '';
  check('the Kotlin twin reads the SAME vector file',
    ktSrc.includes('tests/e2e-ft-hint-vectors.json'));
  check('the Kotlin twin drives the real phone producer',
    ktSrc.includes('FileTransfer.hintFor') && ktSrc.includes('attachHint'));

  const gate = readFileSync(path.join(ROOT, 'tools', 'e2e-gate.mjs'), 'utf8');
  check('the gate declares a MIN_CHECKS floor for this suite',
    /'relay:e2e-ft-hint-contract\.test\.mjs': \d+,/.test(gate));
}

// ── 4. the defect, stated once more as its own named block ────────────────
{
  const row = VECTORS.rows.find((r) => r.id === 'web-offer-hint');
  check('the defect row exists in the vector file', Boolean(row));
  const ft = ftHintFor('FILE_OFFER', row.body);
  const sealed = { e: 1, kid: 'kid-ftA1', s: 42, c: 'c3ViamVjdC1jaXBoZXJ0ZXh0' };
  const withHint = { ...sealed, ft };
  eq('FT-HINT: a browser-sealed FILE_OFFER is ACCEPTED by the relay',
    relay.ftOfferMetadata(withHint).ok, true);
  eq('FT-HINT: and the relay reads the sender\'s own transfer id back',
    relay.ftOfferMetadata(withHint).id, row.body.id);
  // CONTROL: the PRE-FIX shape. If this ever stops being refused, the assertion
  // above has stopped measuring anything — the gate would be open for everyone
  // and this suite would still print a cheerful N/N.
  eq('FT-HINT: the pre-fix BARE envelope is still refused (control)',
    relay.ftOfferMetadata(sealed).ok, false);
  eq('FT-HINT: and that refusal is attributable as bad_hint, not as noise (control)',
    relay.ftOfferMetadata(sealed).sealed, true);
  // CONTROL: the producer must be capable of returning null, or every "emits a
  // hint" assertion above is the truthiness of a value that is always there.
  eq('FT-HINT: the producer returns null for a non-offer frame (control)',
    ftHintFor('FILE_ACCEPT', row.body), null);
  eq('FT-HINT: and null for an offer with a malformed id (control)',
    ftHintFor('FILE_OFFER', { ...row.body, id: 'nope' }), null);
}

console.log(`\n${failed === 0 ? 'PASS' : 'FAIL'}  ${passed} passed, ${failed} failed`);
process.exit(failed === 0 ? 0 : 1);

File: dnkdialer-android/app/src/test/java/com/dnkdialer/companion/E2eFtHintContractTest.kt
package com.dnkdialer.companion

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File

/**
 * FT-A1 sealed-offer hint — the PHONE half of the cross-surface contract
 * (RESUME-PROTOCOL v3.0 RULE 30).
 *
 * ## One file, two implementations
 *
 * `tests/e2e-ft-hint-vectors.json` at the repo root is the SINGLE source of
 * truth. The web lane asserts it in `tests/e2e-ft-hint-contract.test.mjs`
 * (`ftHintFor` -> the envelope `sealOutbound` emits -> the REAL relay accessor
 * `ftOfferMetadata` sliced out of server.js); this file asserts the SAME rows
 * against [FileTransfer.hintFor] and [E2eFrameGate]'s private `attachHint`,
 * which is where the phone actually splices the hint onto a sealed envelope.
 * Same pattern as [E2eKdfVectorsTest] and [E2eSasBlockingContractTest]: a
 * shared vector file is the only thing that can catch the two sides quietly
 * disagreeing, because each side's own tests agree with it by construction.
 *
 * ## The defect this exists for
 *
 * server.js:2361-2392 fails CLOSED when a sealed FILE_OFFER carries no
 * plaintext `ft` hint (drop reason `bad_hint`, server.js:2483). This side has
 * attached one since FT-A1 landed; the BROWSER sealed FILE_OFFER as a bare
 * envelope, so every browser->phone file on any sealed pair was refused —
 * Dennis's "0 bytes, no progress bar", proven on PROD 6d0aa98 by ACCEPT-9.
 * That is a web fix. This file pins the phone shape the web side was told to
 * mirror, so the two producers cannot drift apart again in silence.
 *
 * ## No Kotlin product code changes on this lane
 *
 * Ken's ruling: the F1 lane is web + one Kotlin TEST file. Nothing under
 * `app/src/main` is touched here.
 *
 * Run: `gradlew.bat testDebugUnitTest --tests '*E2eFtHintContractTest*'`
 */
class E2eFtHintContractTest {

    /** Unit-test cwd is `dnkdialer-android/app`, so `../..` is the repo root. */
    private val file = File("../../tests/e2e-ft-hint-vectors.json")

    private fun root(): JsonObject {
        assertTrue(
            "the shared vector file is missing at " + file.absolutePath +
                " — the web lane and this one must read the SAME file",
            file.exists()
        )
        return JsonParser.parseString(file.readText()).asJsonObject
    }

    private fun rows(): List<JsonObject> {
        val r = root()
        assertEquals("vector file version", 1, r.get("version").asInt)
        val rows = r.getAsJsonArray("rows").map { it.asJsonObject }
        // A vectors test whose file lost its rows passes vacuously. It must not.
        assertTrue("the vector file still carries its rows", rows.size >= 10)
        return rows
    }

    /**
     * What THIS surface is expected to emit for a row.
     *
     * `expectHintPhone`, when present, records a MEASURED divergence from the
     * web producer rather than smoothing one over: on `hint-id-uppercase` the
     * web side validates the id against the relay's own `^[0-9a-f]{32}$` and
     * emits nothing, while [FileTransfer.hintFor] passes any non-null String
     * through. Both land on the relay's same `bad_hint` verdict, and the phone
     * mints lowercase ids, so it is not a live defect — but it is a real
     * difference and it is stated in the contract instead of being discovered
     * in an outage. Tightening this side is android PRODUCT code and belongs to
     * the vc64 lane; Ken's ACCEPT-9 ruling keeps this web hotfix out of
     * `app/src/main`.
     */
    private fun expectFor(row: JsonObject) =
        if (row.has("expectHintPhone")) row.get("expectHintPhone") else row.get("expectHint")

    /** The sealed body as [E2eFrameGate.attachHint] sees it: primitives only. */
    private fun fieldsOf(body: JsonObject): Map<String, Any?> {
        val fields = mutableMapOf<String, Any?>()
        for ((k, v) in body.entrySet()) {
            fields[k] = if (v.isJsonPrimitive) {
                val p = v.asJsonPrimitive
                if (p.isNumber) p.asNumber else if (p.isString) p.asString else null
            } else null
        }
        return fields
    }

    /**
     * The REAL private splice, invoked reflectively.
     *
     * Re-typing the splice here would make this a second opinion about the same
     * bytes — the mistake a cross-surface contract exists to prevent. Calling
     * the production method needs no session and no keystore: it takes the
     * sealed body and an already-built envelope string.
     */
    private fun attachHint(gate: E2eFrameGate, type: String, sealedJson: String, envelope: String): String {
        val m = E2eFrameGate::class.java.getDeclaredMethod(
            "attachHint", String::class.java, String::class.java, String::class.java
        )
        m.isAccessible = true
        return m.invoke(gate, type, sealedJson, envelope) as String
    }

    private fun gate(): E2eFrameGate =
        E2eFrameGate(sessionProvider = { null }, latchedProvider = { false })

    // ── 1. the hint key and id shape both surfaces are built on ───────────
    @Test
    fun `the vector file names the key and id shape this side splices`() {
        val r = root()
        assertEquals("the hint key", FileTransfer.HINT_KEY, r.get("hintKey").asString)
        assertEquals("the relay's id shape", "^[0-9a-f]{32}\$", r.get("idPattern").asString)
        assertEquals("FILE_OFFER is the frame the hint belongs to", "FILE_OFFER", FileTransfer.OFFER)
    }

    // ── 2. the phone producer over every row ──────────────────────────────
    @Test
    fun `FileTransfer hintFor agrees with the shared vectors on every row`() {
        val idPattern = Regex(root().get("idPattern").asString)
        var offerRows = 0
        for (row in rows()) {
            val id = row.get("id").asString
            val type = row.get("type").asString
            val expect = expectFor(row)
            val hint = FileTransfer.hintFor(type, fieldsOf(row.getAsJsonObject("body")))

            if (expect == null || expect.isJsonNull) {
                assertNull("$id: the phone producer must emit NO hint", hint)
                continue
            }
            offerRows++
            assertNotNull("$id: the phone producer must emit a hint", hint)
            assertEquals("$id: hint id", expect.asJsonObject.get("id").asString, hint!!["id"] as String)
            assertEquals("$id: hint size", expect.asJsonObject.get("size").asLong, hint["size"] as Long)
            // EVERY row must emit a relay-acceptable id except the ONE that
            // records the divergence, where this side is the looser of the two
            // and the emitted id is exactly what the relay will refuse. Writing
            // it as an equality rather than a relaxed assertTrue keeps both
            // directions measured: if this side were tightened (the vc64 lane),
            // the divergence row goes red here and must be retired from the
            // vector file rather than quietly passing.
            assertEquals(
                "$id: does the emitted id match the shape the relay accepts?",
                !row.has("expectHintPhone"),
                idPattern.matches(hint["id"] as String)
            )
        }
        // A loop that asserted nothing because every row went down the null arm
        // would pass. It must not: the rows this contract exists for are the
        // ones that DO produce a hint.
        assertTrue("at least the offer rows produced a hint", offerRows >= 3)
    }

    // ── 3. the real splice, and what it does NOT put on the wire ──────────
    @Test
    fun `attachHint splices the hint as a sibling of the four authenticated fields`() {
        val g = gate()
        val idPattern = Regex(root().get("idPattern").asString)
        var spliced = 0
        for (row in rows()) {
            val rowId = row.get("id").asString
            val type = row.get("type").asString
            val body = row.getAsJsonObject("body")
            val envelopeVec = row.getAsJsonObject("envelope")
            // The four authenticated fields exactly as E2eEnvelope.Sealed.toJson
            // would have produced them for this row.
            val sealed = "{\"e\":" + envelopeVec.get("e").asInt +
                ",\"kid\":\"" + envelopeVec.get("kid").asString + "\"" +
                ",\"s\":" + envelopeVec.get("s").asInt +
                ",\"c\":\"" + envelopeVec.get("c").asString + "\"}"

            val out = attachHint(g, type, body.toString(), sealed)
            val parsed = JsonParser.parseString(out).asJsonObject

            // The authenticated four survive the splice byte for byte — the peer
            // authenticates those and a re-serialise could reorder them.
            assertTrue("$rowId: the splice produced valid JSON", parsed.has("c"))
            assertEquals("$rowId: e survives", envelopeVec.get("e").asInt, parsed.get("e").asInt)
            assertEquals("$rowId: kid survives", envelopeVec.get("kid").asString, parsed.get("kid").asString)
            assertEquals("$rowId: s survives", envelopeVec.get("s").asInt, parsed.get("s").asInt)
            assertEquals("$rowId: c survives", envelopeVec.get("c").asString, parsed.get("c").asString)

            // Nothing that belongs inside the ciphertext leaves in the clear.
            for (leak in listOf("name", "mime", "sha256", "from")) {
                assertFalse("$rowId: $leak must stay sealed", parsed.has(leak))
            }

            val expect = expectFor(row)
            if (expect == null || expect.isJsonNull) {
                assertFalse(
                    "$rowId: no hint may be spliced onto this frame",
                    parsed.has(FileTransfer.HINT_KEY)
                )
                continue
            }
            spliced++
            assertTrue("$rowId: the hint was spliced", parsed.has(FileTransfer.HINT_KEY))
            val ft = parsed.getAsJsonObject(FileTransfer.HINT_KEY)
            assertEquals("$rowId: spliced id", expect.asJsonObject.get("id").asString, ft.get("id").asString)
            assertEquals("$rowId: spliced size", expect.asJsonObject.get("size").asLong, ft.get("size").asLong)
            assertEquals(
                "$rowId: does the spliced id match the shape the relay accepts?",
                !row.has("expectHintPhone"),
                idPattern.matches(ft.get("id").asString)
            )
            assertEquals("$rowId: the hint carries id and size and nothing else", 2, ft.entrySet().size)
        }
        assertTrue("at least the offer rows were spliced", spliced >= 3)
    }

    // ── 4. the defect, and the controls that prove this file can go red ───
    @Test
    fun `the browser row is the same shape this side emits`() {
        val web = rows().first { it.get("id").asString == "web-offer-hint" }
        val phone = rows().first { it.get("id").asString == "phone-offer-hint" }
        for (row in listOf(web, phone)) {
            val hint = FileTransfer.hintFor("FILE_OFFER", fieldsOf(row.getAsJsonObject("body")))
            assertNotNull("both producers must reach a hint for a valid offer body", hint)
            val wire = row.getAsJsonObject("envelope").getAsJsonObject(FileTransfer.HINT_KEY)
            assertEquals("same id on the wire", wire.get("id").asString, hint!!["id"] as String)
            assertEquals("same size on the wire", wire.get("size").asLong, hint["size"] as Long)
        }
        // The PRE-FIX browser wire shape, recorded in the vectors: a valid body
        // whose envelope carries no hint. This side would never emit it, and the
        // relay refuses it — which is the whole defect.
        val bare = rows().first { it.get("id").asString == "no-hint-bare-envelope" }
        assertTrue("the regression row is marked", bare.get("wireOmitsHint").asBoolean)
        assertFalse(
            "the pre-fix envelope carries no hint",
            bare.getAsJsonObject("envelope").has(FileTransfer.HINT_KEY)
        )
        assertNotNull(
            "and yet the producer must hint that very body — the two cannot both be shipped code",
            FileTransfer.hintFor("FILE_OFFER", fieldsOf(bare.getAsJsonObject("body")))
        )
    }

    @Test
    fun `controls — the producer and the splice are both capable of refusing`() {
        val g = gate()
        val valid = mapOf<String, Any?>("id" to "3f2a91c0d4e84b6798aa10ff5c3b7d20", "size" to 4404019L)
        assertNotNull("CONTROL: the producer hints a valid offer", FileTransfer.hintFor("FILE_OFFER", valid))
        assertNull("CONTROL: never for a non-offer frame", FileTransfer.hintFor("FILE_ACCEPT", valid))
        assertNull(
            "CONTROL: never for an offer with no id",
            FileTransfer.hintFor("FILE_OFFER", mapOf("size" to 1L))
        )
        assertNull(
            "CONTROL: never for an offer with no size",
            FileTransfer.hintFor("FILE_OFFER", mapOf("id" to "3f2a91c0d4e84b6798aa10ff5c3b7d20"))
        )
        // CONTROL: the splice returns the envelope UNTOUCHED when there is
        // nothing to hint. If this ever started appending something, every
        // "no hint on a non-offer" assertion above would be measuring nothing.
        val env = "{\"e\":1,\"kid\":\"k\",\"s\":1,\"c\":\"x\"}"
        assertEquals(
            "CONTROL: a non-offer envelope leaves untouched",
            env,
            attachHint(g, "FILE_ACCEPT", "{\"id\":\"3f2a91c0d4e84b6798aa10ff5c3b7d20\"}", env)
        )
        assertEquals(
            "CONTROL: an unparseable body leaves the envelope untouched",
            env,
            attachHint(g, "FILE_OFFER", "not json", env)
        )
    }
}
